Tools/scripts/freeze_modules.py

```diff
@@ -113,67 +169,168 @@ def parse_frozen_spec(rawspec, knownids=None, section=None):
             pkgname = modname
             def iter_subs():
                 for frozenid, pyfile, ispkg in resolved:
-                    assert not knownids or frozenid not in knownids, (frozenid, rawspec)
+                    assert not knownids or frozenid not in knownids, (frozenid, spec)
                     if pkgname:
                         modname = frozenid.replace(pkgid, pkgname, 1)
                     else:
                         modname = frozenid
                     yield frozenid, pyfile, modname, ispkg, section
             submodules = iter_subs()
 
-    spec = (frozenid, pyfile or None, modname, ispkg, section)
-    return spec, submodules
+    info = (frozenid, pyfile or None, modname, ispkg, section)
+    return info, submodules
 
 
-def parse_frozen_specs(rawspecs=FROZEN):
-    seen = set()
-    for section, _specs in rawspecs:
-        for spec in _parse_frozen_specs(_specs, section, seen):
-            frozenid = spec[0]
-            yield spec
-            seen.add(frozenid)
+#######################################
+# frozen source files
+
+class FrozenSource(namedtuple('FrozenSource', 'id pyfile frozenfile')):
+
+    @classmethod
+    def from_raw(cls, raw, destdir=MODULES_DIR):
+        section = None
+        if isinstance(raw, cls):
+            return raw.resolve_frozen(destdir), section
+        elif isinstance(raw, FrozenSource):
+            raise NotImplementedError(raw)
+
+        # Try FrozenModule.
+        try:
+            source = raw.source
+            section = raw.section
+        except AttributeError:
+            pass
+        else:
+            if not isinstance(source, FrozenSource):
+                raise ValueError(f'bad raw value {raw!r}')
+            return cls.from_raw(source), section
+
+        # It must be a raw spec.
+        frozenid, pyfile, *_, section = mod
+        source = cls.from_id(frozenid, pyfile, destdir)
+        return source, section
+
+    @classmethod
+    def from_id(cls, frozenid, pyfile=None, destdir=MODULES_DIR):
+        if not pyfile:
+            pyfile = os.path.join(STDLIB_DIR, *frozenid.split('.')) + '.py'
+            #assert os.path.exists(pyfile), (frozenid, pyfile)
+        frozenfile = resolve_frozen_file(frozenid, destdir)
+        return cls(frozenid, pyfile, frozenfile)
+
+    @property
+    def frozenid(self):
+        return self.id
+
+    @property
+    def modname(self):
+        if self.pyfile.startswith(STDLIB_DIR):
+            return self.id
+        return None
 
+    @property
+    def symbol(self):
+        # This matches what we do in Programs/_freeze_module.c:
+        name = self.frozenid.replace('.', '_')
+        return '_Py_M__' + name
+
+    def resolve_frozen(self, destdir=MODULES_DIR):
+        if not destdir:
+            return self
```

Comment:
It looks to me like destdir can never be None or empty?

Reply:
This method is no longer used so I've dropped it.